mmv1/products/vertexai/api.yaml

```diff
@@ -433,4 +433,191 @@ objects:
             description: |
               The disk utilization of the MetadataStore in bytes.
             output: true
-
+# Vertex AI Matching Engine Index
+  - !ruby/object:Api::Resource
+    name: Index
+    base_url: projects/{{project}}/locations/{{region}}/indexes
+    create_url: projects/{{project}}/locations/{{region}}/indexes
+    self_link: projects/{{project}}/locations/{{region}}/indexes/{{name}}
+    update_verb: :PATCH
```

Comment:
If you set `update_mask: true` here, that will make it so that only updated fields are included in the PATCH request - it sounds like the PATCH request has to exclude contentsDeltaUri if it wants to update other fields. This should resolve the issue.

You might also need to set [update_mask_fields](https://github.com/GoogleCloudPlatform/magic-modules/blob/5268d03a79ba5644c9b8252175f642245881c8b7/mmv1/products/networkmanagement/terraform.yaml#L42) for all nested fields in order for them to be updated properly... it varies from API to API. I'd recommend testing _without_ update_mask_fields first.

```suggestion
    update_verb: :PATCH
    update_mask: true
```

Reply:
Thank you so much for your clear advice! I'll give it a try with a non-nested field first

Reply:
yes, `update_mask: true` worked for the case updating the description locally. I'll add update_mask_fields later,

Reply:
I tried to update `update_mask_fields` for multi-nested objects. Only `metadata.{isCompleteOverwrite,contentsDeltaUri}` are applied. Therefore, I'll put all fields in `update_mask_fields` of  `metadata`.

<details>
<summary>details</summary>

```diff
diff --git a/mmv1/products/vertexai/terraform.yaml b/mmv1/products/vertexai/terraform.yaml
index 634cc9991..ae1af17c3 100644
--- a/mmv1/products/vertexai/terraform.yaml
+++ b/mmv1/products/vertexai/terraform.yaml
@@ -96,6 +96,20 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         ignore_read: true
       metadata.contentsDeltaUri: !ruby/object:Overrides::Terraform::PropertyOverride
         custom_flatten: templates/terraform/custom_flatten/vertex_ai_index_ignore_contents_delta_uri.go.erb
+      metadata: !ruby/object:Overrides::Terraform::PropertyOverride
+        update_mask_fields:
+          - metadata.contentsDeltaUri
+          - metadata.isCompleteOverwrite
+      metadata.config: !ruby/object:Overrides::Terraform::PropertyOverride
+        update_mask_fields:
+          - metadata.config.dimensions
+          - metadata.config.approximateNeighborsCount
+          - metadata.config.distanceMeasureType
+          - metadata.config.featureNormType
+      metadata.config.algorithmConfig.treeAhConfig: !ruby/object:Overrides::Terraform::PropertyOverride
+        update_mask_fields:
+          - metadata.config.algorithmConfig.treeAhConfig.leafNodeEmbeddingCount
+          - metadata.config.algorithmConfig.treeAhConfig.leafNodesToSearchPercent
   FeaturestoreEntitytype: !ruby/object:Overrides::Terraform::ResourceOverride
     import_format: ["{{%featurestore}}/entityTypes/{{name}}"]
     autogen_async: false
```

```diff
diff --git a/google/resource_vertex_ai_index.go b/google/resource_vertex_ai_index.go
index d61d06d45..3982566cd 100644
--- a/google/resource_vertex_ai_index.go
+++ b/google/resource_vertex_ai_index.go
@@ -21,7 +21,15 @@ import (
 	"strings"
 	"time"
 
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/customdiff"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/logging"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/structure"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
 )
 
 func resourceVertexAIIndex() *schema.Resource {
@@ -494,7 +502,8 @@ func resourceVertexAIIndexUpdate(d *schema.ResourceData, meta interface{}) error
 	}
 
 	if d.HasChange("metadata") {
-		updateMask = append(updateMask, "metadata")
+		updateMask = append(updateMask, "metadata.contentsDeltaUri",
+			"metadata.isCompleteOverwrite")
 	}
 
 	if d.HasChange("labels") {
```
</details>

Reply:
I tested to update m`etadata.config.dimensions` with the above strategy. However, `metadata.config.dimensions` couldn't be updated. I'm wondering if we put `metadata.contentsDeltaUri` in the updateMask query parameter. We can't update the other metadata fields at the same time based on [this quote](https://cloud.google.com/vertex-ai/docs/matching-engine/update-rebuild-index#update_index_content_with_batch_updates):

> Please note that if you set the contentsDeltaUri field when calling IndexService.UpdateIndex, then no other index fields (such as displayName, description, or userLabels) can be also updated as part of the same call.

<details>
<summary>details</summary>

```diff
diff --git a/mmv1/products/vertexai/terraform.yaml b/mmv1/products/vertexai/terraform.yaml
index 634cc9991..25bcf7543 100644
--- a/mmv1/products/vertexai/terraform.yaml
+++ b/mmv1/products/vertexai/terraform.yaml
@@ -96,6 +96,16 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         ignore_read: true
       metadata.contentsDeltaUri: !ruby/object:Overrides::Terraform::PropertyOverride
         custom_flatten: templates/terraform/custom_flatten/vertex_ai_index_ignore_contents_delta_uri.go.erb
+      metadata: !ruby/object:Overrides::Terraform::PropertyOverride
+        update_mask_fields:
+          - metadata.contentsDeltaUri
+          - metadata.isCompleteOverwrite
+          - metadata.config.dimensions
+          - metadata.config.approximateNeighborsCount
+          - metadata.config.distanceMeasureType
+          - metadata.config.featureNormType
+          - metadata.config.algorithmConfig.treeAhConfig.leafNodeEmbeddingCount
+          - metadata.config.algorithmConfig.treeAhConfig.leafNodesToSearchPercent
   FeaturestoreEntitytype: !ruby/object:Overrides::Terraform::ResourceOverride
     import_format: ["{{%featurestore}}/entityTypes/{{name}}"]
     autogen_async: false
diff --git a/mmv1/third_party/terraform/tests/resource_vertex_ai_index_test.go b/mmv1/third_party/terraform/tests/resource_vertex_ai_index_test.go
index f2afdc34c..3cc179047 100644
--- a/mmv1/third_party/terraform/tests/resource_vertex_ai_index_test.go
+++ b/mmv1/third_party/terraform/tests/resource_vertex_ai_index_test.go
@@ -119,7 +119,7 @@ resource "google_vertex_ai_index" "index" {
     contents_delta_uri = "gs://${google_storage_bucket.bucket.name}/contents"
     config {
       dimensions = 2
-      approximate_neighbors_count = 150
+      approximate_neighbors_count = 100
       distance_measure_type = "DOT_PRODUCT_DISTANCE"
       algorithm_config {
         tree_ah_config {
```

```
---[ REQUEST ]---------------------------------------
PATCH /v1/projects/kouzoh-p-kohama/locations/us-central1/indexes/5511570315064377344?alt=json&updateMask=description%2Cmetadata.contentsDeltaUri%2Cmetadata.isCompleteOverwrite%2Cmetadata.config.dimensions%2Cmetadata.config.approximateNeighborsCount%2Cmetadata.config.distanceMeasureType%2Cmetadata.config.featureNormType%2Cmetadata.config.algorithmConfig.treeAhConfig.leafNodeEmbeddingCount%2Cmetadata.config.algorithmConfig.treeAhConfig.leafNodesToSearchPercent HTTP/1.1^M
Host: us-central1-aiplatform.googleapis.com^M
User-Agent: Terraform/1.1.4 (+https://www.terraform.io) Terraform-Plugin-SDK/2.10.1 terraform-provider-google/acc^M
Content-Length: 496^M
Content-Type: application/json^M
Accept-Encoding: gzip^M
^M
{
 "description": "index for test (updated)",
 "displayName": "tf-test-test-indexjpuf16f07i",
 "indexUpdateMethod": "BATCH_UPDATE",
 "labels": {
  "foo": "bar"
 },
 "metadata": {
  "config": {
   "algorithmConfig": {
    "bruteForceConfig": null,
    "treeAhConfig": {
     "leafNodeEmbeddingCount": 500,
     "leafNodesToSearchPercent": 7
    }
   },
   "approximateNeighborsCount": 100,
   "dimensions": 2,
   "distanceMeasureType": "DOT_PRODUCT_DISTANCE",
   "featureNormType": "NONE"
  },
  "contentsDeltaUri": "gs://kouzoh-p-kohama-tf-test-vertex-ai-index-testjpuf16f07i/contents"
 }
}

-----------------------------------------------------
2022/11/19 23:35:06 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK^M
Alt-Svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"^M
Cache-Control: private^M
Content-Type: application/json; charset=UTF-8^M
Date: Sun, 20 Nov 2022 07:35:06 GMT^M
Server: ESF^M
Vary: Origin^M
Vary: X-Origin^M
Vary: Referer^M
X-Content-Type-Options: nosniff^M
X-Frame-Options: SAMEORIGIN^M
X-Xss-Protection: 0^M
^M
{
  "name": "projects/415528280434/locations/us-central1/indexes/5511570315064377344/operations/6200380608602963968",
  "metadata": {
    "@type": "type.googleapis.com/google.cloud.aiplatform.v1.UpdateIndexOperationMetadata",
    "genericMetadata": {
      "createTime": "2022-11-20T07:35:06.726769Z",
      "updateTime": "2022-11-20T07:35:06.726769Z"
    }
  },
  "done": true,
  "response": {
    "@type": "type.googleapis.com/google.cloud.aiplatform.v1.Index",
    "name": "projects/415528280434/locations/us-central1/indexes/5511570315064377344"
  }
}

-----------------------------------------------------
```

```
 | After applying this test step, the plan was not empty.
  | stdout:
  |
  |
  | Terraform used the selected providers to generate the following execution
  | plan. Resource actions are indicated with the following symbols:
  |   ~ update in-place
  |
  | Terraform will perform the following actions:
  |
  |   # google_vertex_ai_index.index will be updated in-place
  |   ~ resource "google_vertex_ai_index" "index" {
  |         id                  = "projects/kouzoh-p-kohama/locations/us-central1/indexes/5511570315064377344"
  |         name                = "5511570315064377344"
  |         # (11 unchanged attributes hidden)
  |
  |       ~ metadata {
  |             # (2 unchanged attributes hidden)
  |
  |           ~ config {
  |               ~ approximate_neighbors_count = 150 -> 100
  |                 # (3 unchanged attributes hidden)
  |
  |                 # (1 unchanged block hidden)
  |             }
  |         }
  |     }
  |
  | Plan: 0 to add, 1 to change, 0 to destroy.
```

</details>

Reply:
I could use granular update masks with the `pre_update` custom code. However, it turns out the patch can't update `metadata.config` and `index_update_method`. 

It seems the patch can use the fields: `contentsDelta`, `isCompleteOverwrite` under the metadata in-place. Because those fields don't appear in a response of get, we need to `ignore_read_extra` for them, though... 

Anyway, I'll push code to use granular update masks. And, I'll put `input: true` on the un-patchable fields.